module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,238 @@
+#include "Director.hpp"
+
+#include "extension/Configuration.hpp"
+
+namespace module {
+    namespace extension {
+
+        using ::extension::Configuration;
+        using ::extension::behaviour::Task;
+        using ::extension::behaviour::commands::CausingExpression;
+        using ::extension::behaviour::commands::DirectorTask;
+        using ::extension::behaviour::commands::ProviderAction;
+        using ::extension::behaviour::commands::ProviderDone;
+        using ::extension::behaviour::commands::ProvidesReaction;
+        using ::extension::behaviour::commands::WhenExpression;
+        using ::NUClear::threading::Reaction;
+        using Unbind = NUClear::dsl::operation::Unbind<ProvidesReaction>;
+
+        using TaskPack = std::vector<std::shared_ptr<const DirectorTask>>;
+
+        /**
+         * This message gets emitted when a state we are monitoring is updated.
+         * When a `When` condition is being waited on a reaction will start monitoring that state.
+         * When it updates we check to see if this would now satisfy the conditions and if it does we emit
+         */
+        struct StateUpdate {
+            StateUpdate(const uint64_t& provider_id, const std::type_index& type, const int& state)
+                : provider_id(provider_id), type(type), state(state) {}
+
+            /// The provider_id which was waiting on this type
+            uint64_t provider_id;
+            /// The enum type that this enum was listening for
+            std::type_index type;
+            /// The new value for the enum
+            int state;
+        };
+
+        /**
+         * Adds a provider to the list of active providers
+         *
+         * @param data_type the data type that this provider services
+         * @param r         the reaction that is used to run this provider
+         * @param action    the action type for this provider (entering, leaving, normal)
+         */
+        void Director::add_provider(const std::type_index& data_type,
+                                    const std::shared_ptr<Reaction>& r,
+                                    const ProviderAction& action) {
+
+            auto& group = groups[data_type];
+
+            // Check if we already inserted this element as a provider
+            if (providers.count(r->id) != 0) {
+                throw std::runtime_error("You cannot have multiple provider DSL words in a single on statement.");
+            }
+
+            auto provider = std::make_shared<Provider>(action, data_type, r);
+            group.providers.emplace_back(provider);
+            providers.emplace(r->id, provider);
+        }
+
+        /**
+         * Removes a provider from our list when it is unbound
```

Comment:
I might be misunderstanding this function 
```suggestion
         * Unbinds a provider and removes it from our list
```

Reply:
The comment that's there is correct. The "Unbind" of a provider emits something which calls this reaction.

So this is a response to being unbound, not an unbinding itself

Reply:
Is it normal to add information about when a function should be called to the function description?